apm-agent-core/src/main/java/co/elastic/apm/agent/bci/HelperClassManager.java

```diff
@@ -128,52 +167,169 @@ public T getForClassLoaderOfClass(Class<?> classOfTargetClassLoader) {
+                }
+                Iterator<WeakReference<List<Object>>> helperImplListIterator = clId2helperImplListMap.values().iterator();
+                while (helperImplListIterator.hasNext()) {
+                    if (helperImplListIterator.next().get() == null) {
+                        helperImplListIterator.remove();
+                    }
+                }
+
+                helper = createHelper(targetCl, tracer, implementation, additionalHelpers);
+
+                List<Object> helperImplList = null;
+                WeakReference<List<Object>> helperImplListRef = clId2helperImplListMap.get(clIdentityHashCode);
+                if (helperImplListRef != null) {
+                    helperImplList = helperImplListRef.get();
                 }
-                // the helper class may have a no-arg or a ElasticApmTracer constructor
-                // this is preferable to a init method,
-                // as it allows the tracer instance variable to be non-null
-                try {
-                    return helperClass.getDeclaredConstructor(ElasticApmTracer.class).newInstance(tracer);
-                } catch (NoSuchMethodException e) {
-                    return helperClass.getDeclaredConstructor().newInstance();
+                if (helperImplList == null) {
+                    Class<?> helperHolderClass = injectClass(targetCl, pd, "co.elastic.apm.agent.bci.HelperHolder", true);
+                    //noinspection unchecked
+                    helperImplList = (List<Object>) helperHolderClass.getField("helperInstanceList").get(null);
+                    clId2helperImplListMap.put(clIdentityHashCode, new WeakReference<>(helperImplList));
                 }
-            } catch (Exception e) {
-                throw new RuntimeException(e);
+
+                helperImplList.add(helper);
+                clId2helperMap.put(clIdentityHashCode, new WeakReference<>(helper));
+            } else {
+                helper = helperRef.get();
+            }
+            return helper;
+        }
+
+        @Override
+        protected void markFailedClassLoader(ClassLoader failedClassLoader) {
+            Integer classLoaderId = System.identityHashCode(failedClassLoader);
+            WeakReference<T> former = clId2helperMap.putIfAbsent(classLoaderId, new WeakReference<>((T) null));
+            if (former == null) {
+                failedClassLoaderSet.putIfAbsent(classLoaderId, Boolean.TRUE);
             }
         }
+    }
 
-        private static List<String> asList(String implementation, String[] additionalHelpers) {
-            final ArrayList<String> list = new ArrayList<>(additionalHelpers.length + 1);
-            list.add(implementation);
-            list.addAll(Arrays.asList(additionalHelpers));
-            return list;
+    static Class injectClass(@Nullable ClassLoader targetClassLoader, ProtectionDomain pd, String className, boolean isBootstrapClass) throws IOException, ClassNotFoundException {
```

Comment:
I'm not sure I got this right. Is my understanding of this correct?
This injects the `HelperHolder` directly into the target CL, right? You do this so that there is a strong reference from the target CL to the helper classes, which are loaded from a child of the target CL. This prevents the helper classes to be collected, because they are otherwise only weakly referenced by from this class, which resides in the system/bootstrap CL.

What I don't quite understand is, when you already inject the `HelperHolder` directly into the target CL, why don't you just inject the helper classes directly into the target CL. Then you also don't need the `HelperHolder` anymore.

Reply:
Because injecting (i.e. defining not as intended) a class exposes you to JVM quirks, like how and when exactly the class and its dependencies are resolved, how loaded classes are searched for when referenced, what effect is there that the JVM only has a "defining loader" to the class and not an "initiating loader" and other things I did not think about. 
All those risks are considerably reduced if you are injecting a minimal class with no dependencies and immediately losing its reference so that it will never be used. This way all the code we run is properly loaded using the JVM mainstream mechanisms.

Reply:
Is it to avoid that the helper classes in the system/bootstrap CL are referenced by making sure to only reference those defined by the `ChildFirstClassLoader`?

Reply:
BTW, what you suggested doesn't solve the problem of the reference- how would I prevent my helper instances from being GCed?

Reply:
This is so that injected classes are never really used or even referenced. Not referencing the incorrect instance is achieved by using a ref to a bootstrap-loaded interface instead of a concrete instance.